src/cpl/share_esmf/ExcessIceStreamType.F90

```diff
@@ -0,0 +1,295 @@
+      ! call get_curr_date(year, mon, day, sec)
+      year = 1996
+      mon  = 12
+      day  = 31
+      sec  = 0
+      mcdate = year*10000 + mon*100 + day
+      
+      call shr_strdata_advance(sdat_exice, ymd=mcdate, tod=sec, logunit=iulog, istr='exice', rc=rc) 
+         if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) then
+            call ESMF_Finalize(endflag=ESMF_END_ABORT)
+         end if
+        
+         ! Get pointer for stream data that is time and spatially interpolate to model time and grid
+         do n = 1,size(stream_varnames)
+          call dshr_fldbun_getFldPtr(sdat_exice%pstrm(1)%fldbun_model, stream_varnames(n), fldptr1=dataptr1d, rc=rc)
+          if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__)) then
+             call ESMF_Finalize(endflag=ESMF_END_ABORT)
+          end if
+          if (trim(stream_varnames(n)) == 'EXICE') then
+            ig = 0
+            do g = bounds%begg,bounds%endg
+               ig = ig+1
+               this%exice_bulk(g) = dataptr1d(ig)
+            end do
+          end if
+         end do
+    end if
+  end subroutine Init
+
+  subroutine InitAllocate(this, bounds)
+    !
+    ! !DESCRIPTION:
+    ! Allocate module variables and data structures
+    !
+    ! !USES:
+    use shr_infnan_mod, only: nan => shr_infnan_nan, assignment(=)
+    !
+    ! !ARGUMENTS:
+    implicit none
+    class(excessicestream_type) :: this
+    type(bounds_type), intent(in) :: bounds
+    !
+    ! !LOCAL VARIABLES:
+    integer  :: begc, endc
+    integer  :: begg, endg
+    !---------------------------------------------------------------------
+
+    begc = bounds%begc; endc = bounds%endc
+    begg = bounds%begg; endg = bounds%endg
+
+
+    allocate(this%exice_bulk(begg:endg))            ;  this%exice_bulk(:)   = nan
+
+  end subroutine InitAllocate
+
+  subroutine CalcExcessIce(this,bounds,exice_bulk_init)
+  
+  ! only transfers grid values to columns
+   use shr_const_mod   , only : SHR_CONST_TKFRZ
+   use landunit_varcon , only : istwet, istsoil, istcrop, istice
+   use column_varcon   , only : icol_road_perv, icol_road_imperv
+   use clm_varcon      , only : denice 
+   use clm_varcon      , only : tfrz
+   use ColumnType      , only : col
+   use LandunitType    , only : lun
+   implicit none
+   class(excessicestream_type)        :: this
+   type(bounds_type),  intent(in)     :: bounds
+   real(r8)         ,  intent(inout)  :: exice_bulk_init(bounds%begc:bounds%endc) 
+   !
+   ! !LOCAL VARIABLES:
+   integer  :: begc, endc
+   integer  :: begg, endg
+   integer  :: c, l, g  !counters
+   
+   exice_bulk_init(bounds%begc:bounds%endc)=0.0_r8
+
+   do c = bounds%begc,bounds%endc
+      g = col%gridcell(c)
+      l = col%landunit(c)
+      if ((.not. lun%lakpoi(l)) .and. (.not. lun%urbpoi(l)) .and. (.not. lun%itype(l) == istwet) .and. (.not. lun%itype(l) == istice)) then  !not lake
```

Comment:
Rather than use an explicit list, we should possibly use a filter here.

Reply:
It's a bit complicated but my logic was to have it here the same way it is `WaterStateType` where it is called from.

Reply:
This would be good to discuss. The nots really mean everything but natural-PFT or crop. So it would probably be more clear to use the soilc filter. With something like:

``` fortran
do fc = 1,num_soilc
   c = filter_soilc(fc)
   g = col%gridcell(c)
   exice_bulk_init(c)=this%exice_bulk(g)
```

Which is much simpler and more easily understood.

You don't really need the else in here, because exice_bulk_init is already being explicitly set to 0.0 everywhere.

Reply:
@mvdebolskiy what part of WaterStateType did you mean? The InitCold part? That is using if statements on all the different types, but it's doing that to make sure it sets something for each type. I'd say this is a little different than that.

Reply:
I meant InitCold. I've put those redundant conditions everywhere, just to ensure that `excess_ice_col` will get 0.0 everywhere except for vegetated and crop units. The filter should be fine, but I was just too cautious.

Reply:
We realized this is just how we do things for InitCold, so this should stay the same. So I'm resolving this conversation.